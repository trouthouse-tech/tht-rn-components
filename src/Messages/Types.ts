import {GestureResponderEvent} from 'react-native';

type BaseProps = {
  name: string;
  date: Date;
  status: MessageStatusEnum;
  image?: string;
};

type BaseTogglerProps = {
  onPress: (selectedMessageStatus: MessageType) => void;
};

export enum MessageStatusEnum {
  Scheduled = 'Scheduled',
  Canceled = 'Canceled',
  Completed = 'Completed',
}

export enum MessageType {
  Sessions = 'Sessions',
  Requests = 'Requests',
}

export enum ScheduledMessageType {
  Scheduled = 'Scheduled',
  History = 'History',
}

//the first component type props
export type MessageItemProps = BaseProps & {
  messageType: MessageType;
};

export type MessageListProps = {
  messages: MessageItemProps[];
};

export type MessageTogglerProps = BaseTogglerProps & {
  currentMessageType: MessageType;
};

//the second component type props
export type ScheduledMessageItemProps = BaseProps & {
  notes: string;
  durationInHours: number;
  messageType: ScheduledMessageType;
};

export type ScheduledMessageListProps = {
  messages: ScheduledMessageItemProps[];
};
