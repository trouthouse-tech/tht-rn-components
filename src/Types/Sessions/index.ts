import {BaseMessageProps} from '../Base/';

export enum SessionMessageTypeEnum {
  Scheduled = 'Scheduled',
  History = 'History',
}

export type SessionMessageItemProps = BaseMessageProps & {
  notes: string;
  durationInHours: number;
};

export type SessionMessageListProps = {
  sessions: SessionMessageItemProps[];
};

export type SessionTogglerProps = {
  currentMessageType: SessionMessageTypeEnum;
  onPress: (selectedMessageStatus: SessionMessageTypeEnum) => void;
};
