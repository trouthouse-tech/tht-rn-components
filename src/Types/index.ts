import * as BaseTypes from './Base/';
import * as MessageTypes from './Messages/';
import * as SessionTypes from './Sessions/';

export {BaseTypes, MessageTypes, SessionTypes};
