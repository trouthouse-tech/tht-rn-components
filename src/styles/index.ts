import * as Colors from './color';
import * as Fonts from './font';

export {Colors, Fonts};
