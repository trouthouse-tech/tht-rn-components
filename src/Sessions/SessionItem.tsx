import React from 'react';
import moment from 'moment';

import {View, Text, Image, StyleSheet, Dimensions} from 'react-native';

import {MessageTypes} from '../Types';
import {Colors, Fonts} from '../styles';

export const SessionItem: React.FC<MessageTypes.SessionMessageItemProps> = (
  props
) => {
  const {
    name,
    date,
    status,
    image = 'https://via.placeholder.com/150',
    notes,
    durationInHours,
  } = props;
  return (
    <View style={styles.messageStyle}>
      <View>
        <Text style={styles.nameTextStyle}>{name}</Text>

        <Text style={styles.notesTextStyle} numberOfLines={2}>
          {notes}
        </Text>

        {status !== MessageTypes.MessageStatusEnum.Completed ? (
          <View style={styles.dateTextViewStyle}>
            <Text style={styles.dateDisplayTextStyle}>
              {moment(date).format('DD/MM/YY')}
            </Text>
            <Text style={styles.dateDisplayTextStyle}>
              {moment(date).format('h:mm a')}
            </Text>
            <Text style={styles.dateDisplayTextStyle}>
              {durationInHours}hr{durationInHours > 1 ? 's' : ''}
            </Text>
          </View>
        ) : (
          <Text style={styles.statusDisplayTextStyle}>{status}</Text>
        )}
      </View>
      <Image source={{uri: image}} style={styles.messageImgStyle} />
    </View>
  );
};

const deviceHeight = Dimensions.get('window').height;
const deviceWidth = Dimensions.get('window').width;

const styles = StyleSheet.create({
  messageStyle: {
    borderWidth: StyleSheet.hairlineWidth,
    borderRadius: 15,
    paddingHorizontal: 20,
    paddingVertical: 10,
    marginVertical: 10,
    justifyContent: 'space-between',
    flex: 1,
    flexDirection: 'row',
    alignContent: 'center',
    alignItems: 'center',
  },
  nameTextStyle: {
    color: Colors.Black,
    fontSize: Fonts.large,
    fontWeight: '700',
  },
  notesTextStyle: {
    fontSize: Fonts.small,
    width: deviceWidth / 3,
  },
  dateTextViewStyle: {
    flexDirection: 'row',
    justifyContent: 'flex-start',
    paddingVertical: 4,
  },
  dateDisplayTextStyle: {
    paddingHorizontal: 3,
    fontSize: Fonts.small,
    fontWeight: '600',
  },
  statusDisplayTextStyle: {
    fontSize: Fonts.small,
    fontWeight: '600',
  },
  messageImgStyle: {
    height: deviceHeight / 7,
    width: deviceHeight / 7,
    borderRadius: 200,
  },
});
