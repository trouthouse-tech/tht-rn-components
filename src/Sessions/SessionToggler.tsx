import React from 'react';

import {View, Text, StyleSheet, TouchableOpacity} from 'react-native';

import {SessionTypes} from '../Types';
import {Colors, Fonts} from '../styles';

const SessionTab: React.FC<SessionTypes.SessionTabProps> = ({
  title,
  isSelected,
  onPress,
}) => {
  return (
    <TouchableOpacity
      onPress={onPress}
      style={[styles.tabStyle, isSelected ? styles.selectedTabStyle : null]}>
      <Text
        style={[
          styles.tabTextStyle,
          isSelected ? styles.selectedTabTextStyle : null,
        ]}>
        {title}
      </Text>
    </TouchableOpacity>
  );
};

export const SessionToggler: React.FC<SessionTypes.SessionTogglerProps> = (
  props
) => {
  const {showScheduled, onPress} = props;

  return (
    <View style={styles.tabsStyle}>
      <SessionTab
        title="Scheduled"
        onPress={() => {
          onPress(true);
        }}
        isSelected={showScheduled}
      />
      <SessionTab
        title="History"
        onPress={() => {
          onPress(false);
        }}
        isSelected={!showScheduled}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  tabsStyle: {
    flexDirection: 'row',
    borderBottomColor: Colors.Grey2,
    borderBottomWidth: StyleSheet.hairlineWidth,
    marginTop: 5,
    marginHorizontal: 5,
  },
  tabStyle: {
    fontSize: Fonts.normal,
    color: Colors.Grey,
    flex: 1,
    alignItems: 'center',
    paddingBottom: 10,
  },
  tabTextStyle: {
    color: Colors.Black,
    fontSize: Fonts.large,
  },
  selectedTabStyle: {
    fontSize: Fonts.normal,
    borderBottomColor: Colors.Black,
    borderBottomWidth: 3,
  },
  selectedTabTextStyle: {
    fontWeight: 'bold',
  },
});
