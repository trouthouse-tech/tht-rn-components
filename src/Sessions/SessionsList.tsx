import React from 'react';

import {View, FlatList, StyleSheet} from 'react-native';

import {MessageTypes} from '../Types';
import {SessionItem} from './SessionItem';

export const SessionList: React.FC<MessageTypes.SessionMessageListProps> = ({
  messages,
}) => {
  return (
    <View style={styles.listStyle}>
      <FlatList
        keyExtractor={(
          item: MessageTypes.SessionMessageItemProps,
          index: number
        ) => index.toString()}
        data={messages}
        renderItem={({item}: {item: MessageTypes.SessionMessageItemProps}) => (
          <SessionItem
            name={item.name}
            status={item.status}
            messageType={item.messageType}
            date={item.date}
            notes={item.notes}
            durationInHours={item.durationInHours}
          />
        )}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  listStyle: {
    marginVertical: 10,
  },
});
