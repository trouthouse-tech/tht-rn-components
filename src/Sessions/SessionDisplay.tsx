import React from 'react';

import {View} from 'react-native';

import {SessionTypes, BaseTypes} from '../Types';
import {SessionList} from './SessionsList';
import {SessionToggler} from './SessionToggler';

export const SessionDisplay: React.FC<SessionTypes.SessionMessageListProps> = ({
  sessions,
}) => {
  const [currentMessageStatus, setcurrentMessageStatus] = React.useState<
    SessionTypes.SessionMessageTypeEnum
  >(SessionTypes.SessionMessageTypeEnum.Scheduled);

  const sessionsToDisplay = () => {
    if (
      currentMessageStatus === SessionTypes.SessionMessageTypeEnum.Scheduled
    ) {
      return sessions.filter(
        (m) => m.status === BaseTypes.MessageStatusEnum.Scheduled
      );
    } else {
      return sessions.filter(
        (m) => m.status !== BaseTypes.MessageStatusEnum.Scheduled
      );
    }
  };

  const toggleSessionStatus = (
    selectedMessageStatus: SessionTypes.SessionMessageTypeEnum
  ) => {
    setcurrentMessageStatus(selectedMessageStatus);
  };

  return (
    <View>
      <SessionToggler
        onPress={(
          selectedMessageStatus: SessionTypes.SessionMessageTypeEnum
        ) => {
          return toggleSessionStatus(selectedMessageStatus);
        }}
        currentMessageType={currentMessageStatus}
      />
      <SessionList sessions={sessionsToDisplay()} />
    </View>
  );
};
