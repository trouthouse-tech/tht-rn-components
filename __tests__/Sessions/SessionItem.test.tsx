import {render, cleanup} from 'react-native-testing-library';
import React from 'react';

import {SessionItem} from '../../src/Sessions/SessionItem';
import {SessionTypes, BaseTypes} from '../../src/Types';

afterEach(cleanup);

describe('<MessageItem />', () => {
  const session: SessionTypes.SessionMessageItemProps = {
    name: 'Edwards Moses',
    status: BaseTypes.MessageStatusEnum.Completed,
    date: 1598457958,
    durationInHours: 1,
    notes: 'Note for test',
  };

  it('should match snapshot', () => {
    const rendered = render(
      <SessionItem
        name={session.name}
        status={session.status}
        date={session.date}
        notes={session.notes}
        durationInHours={session.durationInHours}
      />
    ).toJSON();

    expect(rendered).toMatchSnapshot();
  });
});
