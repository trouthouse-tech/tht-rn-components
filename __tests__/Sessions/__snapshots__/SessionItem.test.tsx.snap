// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageItem /> should match snapshot 1`] = `
<View
  style={
    Object {
      "alignContent": "center",
      "alignItems": "center",
      "borderRadius": 15,
      "borderWidth": 0.5,
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginVertical": 10,
      "paddingHorizontal": 20,
      "paddingVertical": 10,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 20,
          "fontWeight": "700",
        }
      }
    >
      Edwards Moses
    </Text>
    <Text
      numberOfLines={2}
      style={
        Object {
          "fontSize": 16,
          "width": 250,
        }
      }
    >
      Note for test
    </Text>
    <Text
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "600",
        }
      }
    >
      Completed
    </Text>
  </View>
  <Image
    source={
      Object {
        "uri": "https://via.placeholder.com/150",
      }
    }
    style={
      Object {
        "borderRadius": 200,
        "height": 190.57142857142858,
        "width": 190.57142857142858,
      }
    }
  />
</View>
`;
